ntp-proto/src/clock.rs

```diff
@@ -8,5 +10,397 @@ pub trait NtpClock {
+        Self {
+            clock,
+            state: ClockState::StartupBlank,
+            last_update_time: NtpInstant::now(),
+            preferred_poll_interval: PollInterval::MIN,
+            poll_interval_counter: 0,
+            offset: NtpDuration::ZERO,
+        }
+    }
+
+    // Preferred ratio between measured offset
+    // and measurement jitter
+    const POLL_FACTOR: i8 = 4;
+    // Threshold for changing desired poll interval
+    const POLL_ADJUST: i32 = 30;
+
+    pub fn update(
+        &mut self,
+        offset: NtpDuration,
+        jitter: NtpDuration,
+        root_delay: NtpDuration,
+        root_dispersion: NtpDuration,
+        leap_status: NtpLeapIndicator,
+        last_peer_update: NtpInstant,
+    ) -> ClockUpdateResult {
+        // Check that we have a somewhat reasonable result
+        if self.offset_too_large(offset) {
+            return ClockUpdateResult::Panic;
+        }
+
+        // Main decision making
+        //
+        // Combined, this code is responsible for:
+        //  - Filtering large but temporary spikes in the measured
+        //    offset to our timeservers
+        //  - Stepping the clock if a large difference persists long
+        //    enough
+        //  - Ensuring a proper initial frequency measurement on startup
+        //  - Making small (gradual) adjustments to the clock when we
+        //    only have a small error
+        if offset.abs() > NtpDuration::STEP_THRESHOLD {
+            // Large spikes are filtered initialy (to handle weird but temporary network issues)
+            // and then handled by stepping if they persist.
+            match self.state {
+                ClockState::Sync => {
+                    self.state = ClockState::Spike;
+                    return ClockUpdateResult::Ignore;
+                }
+                ClockState::MeasureFreq => {
+                    if NtpInstant::abs_diff(last_peer_update, self.last_update_time)
+                        < NtpDuration::SPIKE_INTERVAL
+                    {
+                        // Initial frequency measurement needs some time
+                        return ClockUpdateResult::Ignore;
+                    }
+
+                    self.set_freq(offset, last_peer_update);
+                    return self.do_step(offset, last_peer_update);
+                }
+                ClockState::Spike => {
+                    if NtpInstant::abs_diff(last_peer_update, self.last_update_time)
+                        < NtpDuration::SPIKE_INTERVAL
+                    {
+                        // Filter out short spikes
+                        return ClockUpdateResult::Ignore;
+                    }
+
+                    // Seems that the large difference reflects reality, since
+                    // it persisted for a significant amount of time. So step
+                    // the clock
+                    return self.do_step(offset, last_peer_update);
+                }
+                _ => {
```

Comment:
I'm personally a fan of making pattern matches exhaustive when you can

Reply:
with an or-pattern in this case, presumably

Reply:
Done